consensus/pacemaker/module.go

```diff
@@ -243,6 +265,87 @@ func (m *pacemaker) NewHeight() {
 		)
 }
 
+// This is called each time there is a NewRound message received by the leader from replicas
+// With the introduction of MinBlockTimeMsec delay, multiple concurrent calls may happen
+// It makes sure that:
+//   - Block proposal is made by only one of the possible `HotstuffLeaderMessageHandler.HandleNewRoundMessage()` concurrent (because delayed) calls
+//   - If the timer expires, the first call to this method will trigger the block proposal
+//   - If a late message is received after a block is proposed by another call, the late message is discarded
+//   - Reads and affectations to pacemaker.latestPrepareRequest state are protected by a mutex
+func (m *pacemaker) ProcessDelayedBlockPrepare(currentHeight uint64) chan bool {
+	m.latestPrepareRequest.m.Lock()
+	defer m.latestPrepareRequest.m.Unlock()
+
+	// Prepare channel for the the current request
+	ch := make(chan bool)
+
+	// First time to build a block for current height, cancel previous timer if any
+	if m.latestPrepareRequest.delayedHeight < currentHeight {
```

Comment:
I don't fully understand the need for `delayedHeight` yet. Can you elaborate (and update the comments)?

There's no way around it?

Reply:
Was resolving an imaginary problem. The idea was that if a timer of an "old" proposal is active and another block proposal wants to setup its own timer; we should discard the old one.

This implies that the same node could to prepare two blocks consecutively without the network deciding on the first one. This should not happen because
* Of that timer (even though it's enforced at the leader level only --discuss this)
* If that node got to propose a block, then there's very low probability that it will be a leader for the next block (highly depends on leader election module)
* The protocol doesn't allow you to propose block h+1 if h is not decided on yet

Yet I'm not sure this could never happen :sweat_smile: